### PR TITLE
daml-lf: move EnvironmentInterface from Scala codegen to interface library

### PR DESCRIPTION
**Pull request by S11001001**
Also move `Interface` and `InterfaceType` out of the `reader` subpackage; they belong with the rest of the data model at the `iface` root.

The specific mechanics of reading a `Dar` all the way to producing an `EnvironmentInterface` are left to Scala codegen's `Codegen` and Java codegen's `CodeGenRunner`; there's no consensus or great stability on the best way to tie these pieces together, but all the pieces might as well be available in the `interface` library at least.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by S11001001** [comment]
Inspired by a query on Slack by @leonelag regarding reading the codegen-relevant parts of dalfs and dars; thanks!

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import data.Ref.Identifier
+
+import scala.collection.breakOut
+import scala.collection.immutable.Map
+import scalaz.syntax.std.map._
+import scalaz.Semigroup
+
+/** The combination of multiple [[Interface]]s, such as from a dar. */
+final case class EnvironmentInterface(typeDecls: Map[Identifier, InterfaceType])
```
Comment (leo-da) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala]: I never liked the name... can we call it `InterfaceUnion`. That is exactly what the combination of multiple `Interface`'s is, right? Or does environment mean anything in this context?

Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala]: "Environment" here means the environment of types available to your program, symbolized `Γ ` in the spec (and most specs for typed languages).

"Union" implies something you cannot do; you cannot union two `Interface`s, or two dalfs for that matter. Do you have a different name suggestion?

Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/EnvironmentInterface.scala]: By the way, the other common pronunciation of `Γ` is "context", which I think is clearly not as good as "environment" in this, er, context. cc @remyhaemmerle-da

**Pull request merged**